nodeInterface/NodeInterface.go

```diff
@@ -590,3 +590,83 @@ func (n NodeInterface) LegacyLookupMessageBatchProof(c ctx, evm mech, batchNum h
 	calldataForL1 = data
 	return
 }
+
+func (n NodeInterface) blockL1Num(l2BlockNum uint64) (uint64, error) {
+	blockHeader, err := n.backend.HeaderByNumber(n.context, rpc.BlockNumber(l2BlockNum))
+	if err != nil {
+		return 0, err
+	}
+	blockL1Num := types.DeserializeHeaderExtraInformation(blockHeader).L1BlockNumber
+	return blockL1Num, nil
+}
+
+func (n NodeInterface) matchL2BlockNumWithL1(l2BlockNum uint64, l1BlockNum uint64) error {
+	blockL1Num, err := n.blockL1Num(l2BlockNum)
+	if err != nil {
+		return fmt.Errorf("failed to get the L1 block number of the L2 block: %v. Error: %w", l2BlockNum, err)
+	}
+	if blockL1Num != l1BlockNum {
+		return fmt.Errorf("no L2 block was found with the given L1 block number. Found L2 block: %v with L1 block number: %v, given L1 block number: %v", l2BlockNum, blockL1Num, l1BlockNum)
+	}
+	return nil
+}
+
+// L2BlockRangeForL1 finds the first and last L2 block numbers that have the given L1 block number
+func (n NodeInterface) L2BlockRangeForL1(c ctx, evm mech, l1BlockNum uint64) (uint64, uint64, error) {
+	currentBlockNum := n.backend.CurrentBlock().Number.Uint64()
+	genesis := n.backend.ChainConfig().ArbitrumChainParams.GenesisBlockNum
+
+	type helperStruct struct {
+		low  uint64
+		high uint64
+	}
+
+	searchHelper := func(currentBlock *helperStruct, fetchedMid *helperStruct, target uint64) error {
+		if currentBlock.low < currentBlock.high {
+			// dont fetch midBlockL1Num if its already fetched above
+			mid := arbmath.SaturatingUAdd(currentBlock.low, currentBlock.high) / 2
+			var midBlockL1Num uint64
+			var err error
+			if mid == fetchedMid.low {
+				midBlockL1Num = fetchedMid.high
+			} else {
+				midBlockL1Num, err = n.blockL1Num(mid)
+				if err != nil {
+					return err
+				}
+				fetchedMid.low = mid
+				fetchedMid.high = midBlockL1Num
+			}
+			if midBlockL1Num < target {
+				currentBlock.low = mid + 1
+			} else {
+				currentBlock.high = mid
+			}
+			return nil
+		}
+		return nil
+	}
+	firstBlock := &helperStruct{low: genesis, high: currentBlockNum}
+	lastBlock := &helperStruct{low: genesis, high: currentBlockNum}
+	// in storedMid low corresponds to value mid and high corresponds to midBlockL1Num inside searchHelper
+	storedMid := &helperStruct{low: currentBlockNum + 1}
+	var err error
```

Comment:
You don't need to define error variable when you are inlining error handling. You can just do

if err := ...  (instead of if err = ..)  below.